build_scripts/colab-tests/validate_colab_dependencies.py

```diff
@@ -0,0 +1,85 @@
+import pathlib
+from poetry.core.semver.version import Version
+
+from collections import defaultdict
+from poetry.factory import Factory
+import re
+
+
+# If you add any new entries to this list, please make sure you test your changes in Colab manually first!
+IGNORED_VIOLATIONS = {
+    "Version check failed for 'click'. Colab has 7.1.2, but 'flask' requires >=8.0",
+    "Version check failed for 'cloudpickle'. Colab has 1.3.0, but 'layer' requires >=2.0.0",
+    "Version check failed for 'grpcio'. Colab has 1.46.3, but 'grpcio-tools' requires >=1.47.0",
+    "Version check failed for 'humanize'. Colab has 0.5.1, but 'layer' requires >=3.11.0",
+    "Version check failed for 'itsdangerous'. Colab has 1.1.0, but 'flask' requires >=2.0",
+    "Version check failed for 'protobuf'. Colab has 3.17.3, but 'layer-api' requires >=3.19.4",
+    "Version check failed for 'pyarrow'. Colab has 6.0.1, but 'layer' requires 7.0.0",
+    "Version check failed for 'urllib3'. Colab has 1.24.3, but 'botocore' requires >=1.25.4,<1.27",
+}
+
+
+def parse_colab_packages(path):
+    with open(path) as packages_file:
+        lines = packages_file.readlines()
+    matches = [re.search(r"^([^\s]*) +(\d+\.\d+.*)$", line) for line in lines]
+
+    versions = {
+        match.group(1): match.group(2) for match in matches if match is not None
+    }
+
+    return versions
+
+
+def get_package_constraints(path):
+    poetry = Factory().create_poetry(path)
+
+    locker = poetry.locker
+
+    constraints = defaultdict(
+        list,
+        {
+            dep.name: [{"dep": dep, "coming_from": "layer"}]
+            for dep in poetry.package.requires
+        },
+    )
+
+    transitive_dependencies = list(
+        locker.get_project_dependency_packages(
+            project_requires=poetry.package.requires, extras=[]
+        )
+    )
+
+    for dependency in transitive_dependencies:
+        for transitive_dependency in dependency.package.requires:
+            constraints[transitive_dependency.name].append(
+                {"dep": transitive_dependency, "coming_from": dependency.name}
+            )
+
+    return constraints
```

Comment:
a bit worried about maintainability of this script if we use plain dicts, with "dep" etc... keys

can we introduce simple dataclasses instead?
maybe use some typing in method signatures just to give more guidance to future contributors